### PR TITLE
Signal lifetime: added test to verify that observe handlers are deallocated

### LINKED ISSUE #2698: Is there any test case to test the lifetime of the observing closures?
For example

signal.observeNext { x in }

The closure { x in } should not be retained after the signal terminates.

### PR DESCRIPTION
**Pull request by NachoSoto**
Brought up by @CosynPa, I realized that it's a pretty important test to have.
Fixes #2698.

The test is a bit verbose, but I wanted to make it very exhaustive.

### CODE REVIEW

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 1]
```diff
@@ -210,6 +210,91 @@ class SignalLifetimeSpec: QuickSpec {
...
+			var observer: Signal<Int, TestError>.Observer!
+
+			var tokenWasDeallocated: Bool!
+
+			var token: Token? = nil
+			weak var weakToken: Token?
+
+			func expectTokenNotDeallocated() {
+				expect(tokenWasDeallocated) == false
+				expect(weakToken).toNot(beNil())
```
Comment (mdiep) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: These two tests seem redundant. Shouldn't it be enough to just test a weak reference? In which case, we could remove `Token` and use an object from Foundation or Swift that supports weak references?

Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Yeah I guess you're right. I wanted to be as exhaustive as possible but I can't think of any reason why just relying on `weak` wouldn't be enough.

Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Changed.

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 2]
```diff
@@ -210,6 +210,71 @@ class SignalLifetimeSpec: QuickSpec {
 				expect(signal).to(beNil())
 			}
 		}
+
+		describe("observe") {
+			var signal: Signal<Int, TestError>!
+			var observer: Signal<Int, TestError>.Observer!
+
+			var tokenWasDeallocated: Bool!
```
Comment (mdiep) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: This is no longer used for anything and can be removed.

**Comment by mdiep** [comment]
Just the one note. This is a great test to have. :clap:

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift (hunk 2 above)
Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: Good catch!

File: ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift [hunk 3]
```diff
@@ -210,6 +210,69 @@ class SignalLifetimeSpec: QuickSpec {
...
+			func expectTokenDeallocated() {
+				expect(weakToken).to(beNil())
+			}
+
+			beforeEach {
+				let (signalTemp, observerTemp) = Signal<Int, TestError>.pipe()
+				signal = signalTemp
+				observer = observerTemp
+
+				tokenWasDeallocated = false
```
Comment (mdiep) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: This is still in here too, which is causing the CI failure. I should have called this out specifically in my last comment—sorry.

Comment (NachoSoto) [inline ReactiveCocoaTests/Swift/SignalLifetimeSpec.swift]: OMG that's what I get for not even compiling my change 😰

**Comment by NachoSoto** [comment]
Bump @mdiep :)

**Comment by mdiep** [comment]
Ha! I never saw the new commit. :relaxed:

**Comment by NachoSoto** [comment]
> Ha! I never saw the new commit. :relaxed:

Yeah `GitHub` is missing that feature! I like that Phabricator does send notifications when that happens.

**Pull request merged**